draft-ietf-quic-transport.md

```diff
@@ -575,47 +627,41 @@ QUIC's connection establishment begins with version negotiation, since all
+client MUST include the version in the packet header.  This allows the server to
+identify early packets and to enable version negotiation.
 
-When the server receives a packet from a client with the VERSION flag set for a
-connection that has not yet been established, it compares the client's version
-to the versions it supports.
+When the server receives a packet from a client with the VERSION flag set, it
+compares the client's version to the versions it supports.
 
-* If the client's version is acceptable to the server, the server MUST use this
-  protocol version for the lifetime of the connection.  All subsequent packets
-  sent by the server MUST have the version flag off.
+If the version selected by the client is not acceptable to the server, the
+server discards the incoming packet and responds with a version negotiation
+packet ({{version-negotiation-packet}}).  This includes the VERSION flag and a
+list of versions that the server will accept.  A server MUST send a version
+negotiation packet for every packet that it receives with an unacceptable
+version.
 
-* If the client's version is not acceptable to the server, the server MUST send
-  a Version Negotiation packet to the client.  This packet will have the VERSION
-  flag set and will include the server's set of supported versions.  On
-  subsequently received packets for the same connection ID with the unacceptable
-  version, the server MUST continue responding with a Version Negotiation
-  packet.
+If the packet contains a version that is acceptable to the server, the server
+proceeds with the handshake ({{handshake}}).  All subsequent packets sent by the
+server MUST have the version flag off.  This commits the server to the version
+that the client selected.
 
 When the client receives a Version Negotiation packet from the server, it should
 select an acceptable protocol version.  If such a version is found, the client
 MUST resend all packets using the new version, and the resent packets MUST use
 new packet numbers.  These packets MUST continue to have the VERSION flag set
 and MUST include the new negotiated protocol version.
 
-The client MUST send its version on all packets until it receives a packet from
-the server with the VERSION flag off.  If version negotiation is successful, the
-client should receive a packet from the server with the VERSION flag off
-indicating the end of version negotiation.  All subsequent packets the client
-sends MUST have the version flag off.
-
-Once the server receives a packet from the client with the VERSION flag off, it
-MUST ignore the VERSION flag in subsequently received packets.
+The client MUST include its selected version on all packets until it starts
+protecting packets with 1-RTT keys.  Only unprotected packets and 0-RTT
+protected packets can include a version.  A client MUST NOT change the version
+it uses unless it is in response to a version negotiation packet from the
+server.
```

Comment:
This isn't specific to TLS 1.3, it's a property that is needed by any crypto protocol that needs to distinguish between the different types of keying material that is in use during the handshake.  There are always unencrypted, replayable-protected, and fully-protected packets with a 0-RTT-capable protocol.  If you accept that it is valuable to distinguish between those three types, you end up with this design.